### ISSUE TITLE
Decimal registry parses unit expression without scaling factor using float instead of decimal

### ISSUE THREAD

**Issue by Flix6x** [hgrecco/pint#1505]
I found the following unexpected difference between float type and decimal type registries:

```
import decimal
import pint

ur = pint.UnitRegistry(non_int_type=decimal.Decimal)

pint.Quantity("m/s").to("km/h")  # <Quantity(3.6, 'kilometer / hour')>
ur.Quantity("m/s").to("km/h")  # fails with TypeError: unsupported operand type(s) for *: 'float' and 'decimal.Decimal'
```
If I add an explicit scaling factor, it works as expected:
```
pint.Quantity("1m/s").to("km/h")  # <Quantity(3.6, 'kilometer / hour')>
ur.Quantity("1m/s").to("km/h")  # <Quantity(3.60000, 'kilometer / hour')>
```

I think that underlying this there's a problem with the parsing of the input string, but I could use some help figuring out where the float is being introduced.

Here's what a relevant test could look like:
```
def test_decimal_magnitude():
    ur = pint.UnitRegistry(non_int_type=decimal.Decimal)

    assert isinstance(ur.Quantity("1m/s").magnitude, decimal.Decimal)
    assert not isinstance(ur.Quantity("m/s").magnitude, float)  # unexpected success (magnitude should not be a float)
    assert isinstance(ur.Quantity("m/s").magnitude, decimal.Decimal)  # unexpected fail (magnitude should be a decimal)
```

Both `pint==0.18` and `pint==0.19.1` have this problem.

**Comment by Flix6x** [comment]
`pint==0.20.1` still has this problem. Does anyone know where the float is introduced? I'd love to solve this issue.

To clarify:
```
>>> ur.Quantity("m/s").magnitude
1.0
>>> type(ur.Quantity("m/s").magnitude)
float
```
Above, the float is unexpected, given that we set up the `ur` unit registry to work with decimals. Adding an explicit magnitude works as expected:
```
>>> type(ur.Quantity("1m/s").magnitude)
decimal.Decimal
>>> type(ur.Quantity("1.0m/s").magnitude)
decimal.Decimal
```

**Comment by dalito** [comment]
I wonder if the root of the problem is the same as here https://github.com/hgrecco/pint/issues/1735

**Comment by Flix6x** [comment]
I tracked the issue to [`build_eval_tree(gen).evaluate`](https://github.com/hgrecco/pint/blob/5cbebb618ea8364902512d0bb352d98d149da161/pint/facets/plain/registry.py#L1261), but beyond that point the code becomes too obscure for me. `lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)` evaluates the inputs to `evaluate` correctly, but the output of the call to `evaluate` has the wrong type.

**Comment by hgrecco** [comment]
There are a few offending lines (basically every time a `1` appears that is not written as `self.non_int_type("1")`  or similar):

https://github.com/hgrecco/pint/blob/c208b0edb87b244c4c40d4beba7be8ec32e57eaf/pint/facets/plain/registry.py#L1140

https://github.com/hgrecco/pint/blob/c208b0edb87b244c4c40d4beba7be8ec32e57eaf/pint/facets/plain/registry.py#L1149

https://github.com/hgrecco/pint/blob/c208b0edb87b244c4c40d4beba7be8ec32e57eaf/pint/facets/plain/registry.py#L1257

**Comment by hgrecco** [comment]
Do not celebrate too quickly @Flix6x 

Still checking if this is a good idea. A problem that I see is that we are now forcing non_int_type (Decimal in your case) where we used to insert an integer before. I am reopening until convinced that there are no major problems.

**Comment by Flix6x** [comment]
Well I am still happy being unblocked. :) I hope I can make some other contributions soon with respect to decimal support, now that I got to understand more of the inner logic of Pint.

**Comment by hgrecco** [comment]
I tracked the actual problem (My previous PR worked for this but breaks other things).

Briefly, there are two internal scalar magnitude types in pint: int and non-ints. The first are always of Python's `int` type. The second is user configurable via the `non_int_type` argument and defaults to Python's `float`. When no magnitude was given, the value `1 (type int)` was assumed before my latest (and maybe wrong commit). The reason is that int are arbitrary precision in Python interact well with most other types (including decimal). 

In a series of PRs we made sure that floats do not resurface (for example, we make sure that parsing the definition file honors the selected `non_int_type`). But we is one important part: division

`m/s` is interpreted as `1 m / 1 s` which ends up doing `1 (int) /1 (int)` which in Python returns `1.0 (type float)`. This is the root of this problem.

In the latest changes, we return a `1 (type non_int_type)`. This solves this issue but breaks other. Namely, the only way to make an  Quantity with int type is to do it explicitly `Quantity(1, "meter')` as `Quantity("1 meter")` is parsed as `Quantity(non_int_type("1"), "meter")`

I see three ways out of this conundrum:
1. keep the current behavior. Only explicit ints are ints
2. change the parser to force affinity between units i.e. `1 m/s` is `1 (m/s)` which is `1 * Unit("meter") / Unit("second")`
3. go back to the old behavior and fix division to force non_int_types for ints before dividing.

**Comment by cgevans** [comment]
Option 3—forcing non_int_type for division—ultimately seems like the best option here.  Many of the non_int_type issues with Pint involve either unintended conversions to floats (usually from division), or unintended float-decimal operations failing.  Ensuring that division did not result in a float for a non_int_type registry would not just fix this issue, but also likely prevent other situations where floats might appear unintentionally.

**Comment by hgrecco** [comment]
I agree with you @cgevans. I will benchmark and see.

**Comment by hgrecco** [comment]
Ok. I have just pushed a fix in the way division is handled. It looks good but I will test it and benchmark it further in the next days. If I am not convinced, I will roll back this change and postpone it for the 0.23
